pysetup/mk_to_spec.py

```diff
@@ -0,0 +1,620 @@
+import ast
+import json
+from pathlib import Path
+import string
+from typing import Dict, Optional, Tuple, Iterator
+import re
+from functools import lru_cache
+
+
+from marko.block import BlankLine, Heading, FencedCode, HTMLBlock, Document
+from marko.element import Element
+from marko.ext.gfm import gfm
+from marko.ext.gfm.elements import Table, TableRow, TableCell
+from marko.inline import CodeSpan
+
+from .typing import ProtocolDefinition, VariableDefinition, SpecObject
+
+class MarkdownToSpec:
+    def __init__(self, file_name: Path, preset: Dict[str, str], config: Dict[str, str], preset_name: str):
+        self.preset = preset
+        self.config = config
+        self.preset_name = preset_name
+
+        # Use a single dict to hold all SpecObject fields
+        self.spec = {
+            "functions": {},
+            "protocols": {},
+            "custom_types": {},
+            "preset_dep_custom_types": {},
+            "constant_vars": {},
+            "preset_dep_constant_vars": {},
+            "preset_vars": {},
+            "config_vars": {},
+            "ssz_dep_constants": {},
+            "func_dep_presets": {},
+            "ssz_objects": {},
+            "dataclasses": {},
+        }
+
+        self.all_custom_types: Dict[str, str] = {}
+
+        self.document_iterator: Iterator[Element] = self._parse_document(file_name)
+        self.current_heading_name: str | None = None
```

Comment:
Maybe group these at the top & remove unnecessary blank lines?

Reply:
Do you mean there should not be blank lines at all or that there are too many?

Reply:
@jtraglia do you mean this specific blank line or any blank line inside a method? Like it can be very subjecting how much vertical spacing is correct

Reply:
I meant lines 40, 42, and 43 could be moved up and grouped with the others. No worries, it can stay how it is.